File: java/queue/Queue.java
public class Queue {

	private static class Node {

		private int data;
		private Node next;
		private Node(int data) {
			this.data = data;
		}
	}

	private Node head;
	private Node tail;

	public boolean isEmpty() {
		return head == null;
	}

	public int peek() {
		return (head != null) ? head.data : tail.data;
	}

	public void add(int data) {

		Node node = new Node(data);
		if(head == null) {
			  head = node;
			  return;
		}

		if(tail != null) {
			tail.next = node;
			tail = node;
		}
	}

	public int remove() {

		int data = head.data;
		head = head.next;
		if(head == null) {
			tail = null;
		}

		return data;
	}
}

File: c/queue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct data {
	int data;
	struct data* next;
} data;

typedef struct Queue {
	data* head;
	data* tail;
} Queue;

void add(Queue** queue, int value) {
    data* new = malloc(sizeof(data));
	new->data = value;
	new->next = NULL;

	if((*queue)->head == NULL) {
		(*queue)->head = new;
		(*queue)->tail = new;
	}
	else {
		(*queue)->tail->next = new;
		(*queue)->tail = new;
	}
}

void printQueue(Queue* queue) {
    data* aux = queue->head;
    while(aux) {
        printf("%d ", aux->data);
        aux = aux->next;
    }
    putchar('\n');
}

Queue* createQueue() {
    Queue* new = malloc(sizeof(Queue));
    new->head = NULL;
    new->tail = NULL;

    return new;
}

int main() {
    Queue* fila = createQueue();

    printQueue(fila);

    add(&fila, 16);
    add(&fila, 6);
    add(&fila, 23);

    printQueue(fila);

    return 0;
}